Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/data/MIDIDataType.java

```diff
@@ -0,0 +1,173 @@
+/* ###
+ * IP: GHIDRA
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ghidra.program.model.data;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+
+import ghidra.docking.settings.Settings;
+import ghidra.program.model.mem.MemBuffer;
+import ghidra.program.model.mem.Memory;
+import ghidra.util.Msg;
+
+public class MIDIDataType extends BuiltIn implements Dynamic {
+	public MIDIDataType() {
+		this(null);
+	}
+
+	public MIDIDataType(DataTypeManager dtm) {
+		super(null, "MIDI-Score", dtm);
+	}
+
+	@Override
+	public int getLength() {
+		return -1;
+	}
+
+	@Override
+	public int getLength(MemBuffer buf, int maxLength) {
+		try {
+			return computeLength(buf, maxLength);
+		}
+		catch (Exception e) {
+			Msg.debug(this, "Invalid MIDI data at " + buf.getAddress());
+		}
+		return -1;
+	}
+
+	private long readUnsignedInteger(InputStream stream) throws IOException {
+		long value = 0;
+		for (int index = 0; index < 4; index++) {
+			int currentByte = stream.read();
+			if (currentByte == -1) {
+				throw new EOFException();
+			}
+			value = (value << 8) | currentByte;
+		}
+		return value;
+	}
+	
+	private int readUnsignedShort(InputStream stream) throws IOException {
+		int value = 0;
+		for (int index = 0; index < 2; index++) {
+			int currentByte = stream.read();
+			if (currentByte == -1) {
+				throw new EOFException();
+			}
+			value = (value << 8) | currentByte;
+		}
+		return value;
+	}
+	
+	private int computeLength(MemBuffer buf, int maxLength) throws IOException, InvalidDataTypeException {
+		int computedLength = -1;
+		
+		try (InputStream stream = buf.getInputStream(0, maxLength > 0 ? maxLength : Integer.MAX_VALUE)) {
```

Comment:
`MemBufferByteProvider` paired with `BinaryReader` might work better here as `BinaryReader` already has `readUnsignedShort`, `readNextUnsignedShort`, `readUnsignedInt`, `readNextUnsignedInt`, etc methods. You won't need to worry about `maxLength` or closing it afterwards (there is no close or any need to).